### PR TITLE
Feature pid controller

### PR DESCRIPTION
**Pull request by AE173**

### CODE REVIEW

File: src/PID/PIDController.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package PID;
+
+public interface PIDController {
+  PIDFTerms getPID();
```
Comment (SaarYogev) [inline src/PID/PIDController.java:4]: As the class name suggests, it should be:
``` java
getPIDFTerms()
```

File: src/PID/MotorControllerEnhanced.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
+package PID;
+
+import com.ctre.phoenix.motorcontrol.ControlMode;
+import com.ctre.phoenix.motorcontrol.IMotorControllerEnhanced;
+
+public class MotorControllerEnhanced implements PIDController {
```
Comment (SaarYogev) [inline src/PID/MotorControllerEnhanced.java:6]: ClassName should include PIDController:
`MotorControllerEnhancedPIDController`

File: src/PID/MotorControllerEnhanced.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  private IMotorControllerEnhanced motorControllerEnhanced;
+  private PIDFTerms pidfTerms;
+
+  public MotorControllerEnhanced(IMotorControllerEnhanced motorControllerEnhanced) {
```
Comment (SaarYogev) [inline src/PID/MotorControllerEnhanced.java:10]: I see no reason to create a PID controller with no setpoint. So include it in all constructors.

File: src/PID/MotorControllerEnhanced.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+
+  @Override
+  public PIDFTerms getPID() {
+    return this.pidfTerms;
+  }
+
+  @Override
+  public double getSetpoint() {
+    return motorControllerEnhanced.getClosedLoopTarget(0);
```
Comment (SaarYogev) [inline src/PID/MotorControllerEnhanced.java:38]: Add a field for the PID slot. We might use another one someday 🤷‍♂️ . Of course also make a constructor setting it to 0, so we won't need to bother setting it unless we want to.

File: src/PID/MotorControllerEnhanced.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+      return motorControllerEnhanced.getSelectedSensorVelocity(0);
+    } else if (motorControllerEnhanced.getControlMode() == ControlMode.Position) {
+      return motorControllerEnhanced.getSelectedSensorPosition(0);
+    }
+    return -1;
+  }
+
+  @Override
+  public double getLastError() {
+    return motorControllerEnhanced.getLastError().value;
```
Comment (SaarYogev) [inline src/PID/MotorControllerEnhanced.java:58]: That's not the error you're looking for 😶 

> Gets the last error generated by this object. Not all functions return an error code but can potentially report errors. This function can be used to retrieve those error codes.

You're looking for `getClosedLoopError`.

File: src/PID/MotorControllerEnhancedPIDController.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
+package PID;
+
+import com.ctre.phoenix.motorcontrol.ControlMode;
+import com.ctre.phoenix.motorcontrol.IMotorControllerEnhanced;
+
+public class MotorControllerEnhancedPIDController implements PIDController {
+  private IMotorControllerEnhanced motorControllerEnhanced;
+  private PIDFTerms pidfTerms;
+  private int PIDSlot;
```
Comment (SaarYogev) [inline src/PID/MotorControllerEnhancedPIDController.java:9]: think it should be pidSlot according to conventions? put pid as an abbreviation and intelliJ should tell you.

File: src/pid/PIDControllerRunner.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    this.intervalBetweenMeasurements = INTERVAL_BETWEEN_MEASUREMENTS;
+  }
+
+  public void startPIDLoop() {
+    motorControllerEnhancedPIDController.resetSumOfErrors();
+    motorControllerEnhanced.set(motorControllerEnhanced.getControlMode(),
+        motorControllerEnhancedPIDController.calculate(intervalBetweenMeasurements));
+  }
+
+  public void stopPIDLoop(double speedAfterStop) {
```
Comment (SaarYogev) [inline src/pid/PIDControllerRunner.java:34]: Why would you need to set a specific speed. Like, not the speed it was at nor 0?

File: src/PID/MotorControllerEnhancedPIDController.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    motorControllerEnhanced.config_kD(pidSlot, kD, 100);
+    motorControllerEnhanced.config_kF(pidSlot, kF, 100);
+    this.pidfTerms.setP(kP);
+    this.pidfTerms.setP(kI);
+    this.pidfTerms.setP(kD);
+    this.pidfTerms.setP(kF);
+  }
+
+  public boolean isOnTarget(double tolerance){
+    return ((this.getSetpoint() - Math.abs(this.getCurrentError())) > (this.getSetpoint() - tolerance));
```
[the remark this answers is not in the document]
Comment (AE173) [reply src/PID/MotorControllerEnhancedPIDController.java]: So should I add a function that does that or change this function accordingly? I think changing this should be better as your explanation convinces me there is no real use in knowing if we are currently in the range

File: src/pid/PIDRunner.java [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package pid;
+
+public interface PIDRunner {
+  void startPIDLoop();
+
+  void stopPIDLoop(int remainOrStop);
```
Comment (SaarYogev) [inline src/pid/PIDRunner.java:6]: Should be a boolean, something like `shouldStop`

File: src/pid/PIDControllerRunner.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+  public void startPIDLoop() {
+    this.motorControllerEnhancedPIDRunner.setIntegralAccumulator(0,
+        this.motorControllerEnhancedPIDController.getPidSlot(), 0);
+    motorControllerEnhancedPIDRunner.set(motorControllerEnhancedPIDRunner.getControlMode(),
+        motorControllerEnhancedPIDController.calculate());
+  }
+
+  public void stopPIDLoop(int remainOrStop) {
+    if (remainOrStop == 0)
```
Comment (SaarYogev) [inline src/pid/PIDControllerRunner.java:24]: Always use braces!

File: src/pid/PIDControllerRunner.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+  public PIDControllerRunner(IMotorControllerEnhanced MotorControllerEnhancedPIDRunner,
+                             MotorControllerEnhancedPIDController motorControllerEnhancedPIDController) {
+    this.motorControllerEnhancedPIDRunner = MotorControllerEnhancedPIDRunner;
+    this.motorControllerEnhancedPIDController = motorControllerEnhancedPIDController;
+  }
+
+  public void startPIDLoop() {
+    this.motorControllerEnhancedPIDRunner.setIntegralAccumulator(0,
+        this.motorControllerEnhancedPIDController.getPidSlot(), 0);
```
Comment (SaarYogev) [inline src/pid/PIDControllerRunner.java:18]: Use `TIMEOUT`.

File: src/PID/MotorControllerEnhancedPIDController.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (SaarYogev) [reply src/PID/MotorControllerEnhancedPIDController.java]: As I said, There is definitely a use, with contentious PID, like on a forklift. As we've talked about, this should be **its own interface and class, independent from PID**. Two implementations, and both decoupled from PID, just looking at error, error change (derivative) and targets.

File: src/PID/MotorControllerEnhancedPIDController.java [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+  }
+
+  public boolean isOnTarget(double belowTolerance, double aboveTolerance){
+    return ((this.getSetpoint() - this.getCurrentError()) > (this.getSetpoint() - belowTolerance)) &&
+        ((this.getSetpoint() + this.getCurrentError()) < (this.getSetpoint() + aboveTolerance));
+  }
+
+  public void resetSumOfErrors(){
+    this.sumOfErrors = 0;
```
[the remark this answers is not in the document]
Comment (SaarYogev) [reply src/PID/MotorControllerEnhancedPIDController.java]: That means another method inn PIDController

File: src/pid/PIDRunner.java (hunk 9 above)
Comment (AE173) [reply src/pid/PIDRunner.java:6]: this function is supposed to be in the MotorControllerEnhancedPIDController class no?
`

File: src/PID/MotorControllerEnhancedPIDController.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (AE173) [reply src/PID/MotorControllerEnhancedPIDController.java]: I can't find a function that restarts the previousError

File: src/pid/PIDControllerRunner.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
+package pid;
+
+import com.ctre.phoenix.motorcontrol.ControlMode;
+import com.ctre.phoenix.motorcontrol.IMotorControllerEnhanced;
+
+public class PIDControllerRunner implements PIDRunner {
+  private IMotorControllerEnhanced motorControllerEnhancedPIDRunner;
+  private MotorControllerEnhancedPIDController motorControllerEnhancedPIDController;
```
Comment (SaarYogev) [inline src/pid/PIDControllerRunner.java:8]: Actually really like you put the implementation here, as it's not sensible to use any other with this runner, so this is better in terms of type safety. Good job!

File: src/pid/PIDControllerRunner.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
+package pid;
+
+import com.ctre.phoenix.motorcontrol.ControlMode;
+import com.ctre.phoenix.motorcontrol.IMotorControllerEnhanced;
+
+public class PIDControllerRunner implements PIDRunner {
```
Comment (SaarYogev) [inline src/pid/PIDControllerRunner.java:6]: Add `MotorControllerEnhanched` to this name too.

File: src/pid/MotorControllerEnhancedPIDController.java [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+  }
+
+  public double calculate() {
+    return pidfTerms.getKp() * this.getCurrentError() + pidfTerms.getKi() *
+        this.motorControllerEnhanced.getIntegralAccumulator(this.pidSlot) + pidfTerms.getKd() *
+        this.motorControllerEnhanced.getErrorDerivative(this.pidSlot) + this.pidfTerms.getKf();
+    // p*error+i*sumOfErrors+d*(error-previousError)/interval between measurements + f
+  }
+
+  public void resetAccumulator() {
```
Comment (SaarYogev) [inline src/pid/MotorControllerEnhancedPIDController.java:108]: Change the name to something more generic like `restartControllerState`, as you might be resetting just i, maybe i and d, the important thing is that the controller should be as "fresh" as possible, without state.

File: src/pid/PIDControllerRunner.java [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+    this.motorControllerEnhancedPIDController = motorControllerEnhancedPIDController;
+  }
+
+  public void startPIDLoop() {
+    motorControllerEnhancedPIDController.resetAccumulator();
+    motorControllerEnhancedPIDRunner.set(motorControllerEnhancedPIDRunner.getControlMode(),
+        motorControllerEnhancedPIDController.calculate());
+  }
+
+  public void terminatePIDLoop(int remainOrStop) {
```
Comment (SaarYogev) [inline src/pid/PIDControllerRunner.java:22]: Can't see why you've changed it from stop to terminate (besides sounding cooler 🆒 ).  Start/stop are opposites and it helps understanding that those functions are opposites too.

File: src/pid/MotorControllerEnhancedPIDController.java [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+    this.motorControllerEnhanced = motorControllerEnhanced;
+    this.pidfTerms = new PIDFTerms(0, 0, 0, 0);
+    this.setSetpoint(setpoint);
+    this.pidSlot = pidSlot;
+  }
+
+  public MotorControllerEnhancedPIDController(IMotorControllerEnhanced motorControllerEnhanced, double kP, double kI,
+                                              double kD, double kF, double setpoint) {
+    this.motorControllerEnhanced = motorControllerEnhanced;
+    PIDFTerms pidfTerms = new PIDFTerms(kP, kI, kD, kF);
```
Comment (SaarYogev) [inline src/pid/MotorControllerEnhancedPIDController.java:31]: Line is redundant, you're not usoing the variable.

File: src/pid/MotorControllerEnhancedPIDController.java [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+
+  public boolean isOnTarget(double tolerance) {
+    return Math.abs(getCurrentError()) < tolerance;
+  }
+
+  public boolean isOnTarget(double belowTolerance, double aboveTolerance) {
+    return getCurrentError() > belowTolerance && getCurrentError() < aboveTolerance;
+  }
+
+  public double calculate() {
```
Comment (SaarYogev) [inline src/pid/MotorControllerEnhancedPIDController.java:101]: Like every method here, should be on the interface. You haven't used helping private methods (not that they were needed, those functions are really short as they are).

File: src/pid/PIDControllerRunner.java (hunk 16 above)
Comment (AE173) [reply src/pid/PIDControllerRunner.java:22]: [https://www.google.com/search?q=sad+face&rlz=1C1EKKP_enIL818IL818&oq=sad+face&aqs=chrome..69i57j0l7.2123j0j7&sourceid=chrome&ie=UTF-8](url)

Comment (SaarYogev) [reply src/pid/PIDControllerRunner.java:22]: page no found, but I assume from url it's a 😢

File: src/pid/CtrePIDController.java [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+    return getCurrentError() > belowTolerance && getCurrentError() < aboveTolerance;
+  }
+
+  public void resetAccumulator() {
+    this.ctreMotorController.setIntegralAccumulator(0,
+        this.pidSlot, this.timeoutMs);
+  }
+
+  @Override
+  public void enable(CustomizeControlMode controlMode) {
```
Comment (SaarYogev) [inline src/pid/CtrePIDController.java:54]: It would be nice to get  a parameterless `enable()` function that's configured using ctor parameters. That way you could use the `Controller` interface for other controller types as well (right now it won't be practical since you can't enable them), meaning you could actually use your interface to abstract the controller implementation.

File: src/pid/AbstractController.java [hunk 20]
```diff
@@ -0,0 +1,34 @@
+package pid;
+
+import pid.interfaces.Controller;
+
+public abstract class AbstractController implements Controller {
+  protected PIDFTerms pidfTerms;
+  protected double setpoint;
+
+  public AbstractController(double kP, double kI, double kD, double kF) {
```
Comment (SaarYogev) [inline src/pid/AbstractController.java:9]: You can add a ctor that gets `PIDFTerms` as a parameter and call it from here. That would also enable you to call it from inheritors. Genrally, I think they should create the PIDFTerms outside to make them more meaningful, and to avoid duplicate naming (`driveTrainTerms` instead of `driveTrainP`, `driveTrainI`, `driveTrainD`, `driveTrainF`).

File: src/pid/CtreController.java [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+import sensors.counter.CtreEncoder;
+
+public abstract class CtreController extends AbstractController {
+  protected IMotorControllerEnhanced ctreMotorController;
+  protected CtreEncoder ctreEncoder;
+  protected int pidSlot;
+  protected int timeoutMs;
+
+  public CtreController(IMotorControllerEnhanced motorControllerEnhanced, CtreEncoder ctreEncoder) {
+    super(0, 0, 0, 0);
```
Comment (SaarYogev) [inline src/pid/CtreController.java:19]: There's no reason to pass empty terms

File: src/pid/CtreController.java [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+  @Override
+  public void setPIDFTerms(double kP, double kI, double kD, double kF) {
+    super.setPIDFTerms(kP, kI, kD, kF);
+    ctreMotorController.config_kP(pidSlot, kP, this.timeoutMs);
+    ctreMotorController.config_kI(pidSlot, kI, this.timeoutMs);
+    ctreMotorController.config_kD(pidSlot, kD, this.timeoutMs);
+    ctreMotorController.config_kF(pidSlot, kF, this.timeoutMs);
+  }
+
+  public void setRampSec(double rampSec) {
```
Comment (SaarYogev) [inline src/pid/CtreController.java:77]: Remember, you won't be able to use those through the interface. Also, you probably don't want to change them while the controller is already active. So these things belong in the consructor.

Comment (SaarYogev) [reply src/pid/CtreController.java:77]: Do try to put default values to not make overwhelming though.

Comment (SaarYogev) [reply src/pid/CtreController.java:77]: Also, you can still keep the function itself for complex set functions like `setCurrentLimits`, just make them private and call them from the ctor only.

File: src/pid/CtreMotionMagicController.java [hunk 23]
```diff
@@ -0,0 +1,31 @@
+package pid;
```
Comment (SaarYogev) [inline src/pid/CtreMotionMagicController.java:1]: Move this to another PR.

File: src/pid/CtrePIDController.java [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    return -1;
+  }
+
+  @Override
+  public double getCurrentError() {
+    return this.ctreMotorController.getClosedLoopError(pidSlot);
+  }
+
+  public boolean isOnTarget(double tolerance) {
```
Comment (SaarYogev) [inline src/pid/CtrePIDController.java:40]: `isOnTarget` should move to the `PIDController` interface, and `tolerance` should be fixed in the ctor.

File: src/pid/CtrePIDController.java [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+    return Math.abs(getCurrentError()) < tolerance;
+  }
+
+  public boolean isOnTarget(double belowTolerance, double aboveTolerance) {
+    return getCurrentError() > belowTolerance && getCurrentError() < aboveTolerance;
+  }
+
+  public void resetAccumulator() {
```
Comment (SaarYogev) [inline src/pid/CtrePIDController.java:48]: This too should be in the interface. As I said before, also naming this `restartControllerState` or just `resetState` will make this more open to other implementations (like reseting *kD* to avoid spike in the first iteration after reset).

File: src/pid/CustomizeControlMode.java [hunk 26]
```diff
@@ -0,0 +1,7 @@
+package pid;
+
+public enum CustomizeControlMode {
```
Comment (SaarYogev) [inline src/pid/CustomizeControlMode.java:3]: Customize is a verb, enums should be a noun, Just `ControlMode` would suffice, but since it's taken I'd go for `PIDControlMode`.

File: src/pid/PIDConstants.java [hunk 27]
```diff
@@ -0,0 +1,5 @@
+package pid;
+
+public class PIDConstants {
+  final static int TIMEOUT = 20;
```
Comment (SaarYogev) [inline src/pid/PIDConstants.java:4]: Call this something more specific to CTRE, as this can get confused easily with the PID running interval. Maybe `CTRE_DEVICE_CALLS_TIMEOUT`?

File: src/sensors/counter/DistanceMeterEncoder.java [hunk 28]
```diff
@@ -19,6 +19,11 @@ public double getRate() {
         return counter.getCount();
     }
 
+    @Override
+    public void setCount(int count) {
```
Comment (SaarYogev) [inline src/sensors/counter/DistanceMeterEncoder.java:23]: You even might be able to see in some encoder interfaces that adding a `set` does not fit the FIRST interfaces.

**Review by SaarYogev — CHANGES_REQUESTED** [review]
Here are my changes. remember to go over these after you split the PRs to PID, MotionMagic and MotionProfile.

File: src/pid/CtreController.java (hunk 22 above)
Comment (AE173) [reply src/pid/CtreController.java:77]: Why can't the users just call these methods before they call enable()? Too many constructors, in my opinion, because there are not really any default values that will work in every subsystem

File: src/pid/CtrePIDController.java (hunk 24 above)
Comment (AE173) [reply src/pid/CtrePIDController.java:40]: But then what about aboveTolerance and belowTolerance in the other isOnTarget? Will that function receive two parameters?

File: src/pid/CtreController.java (hunk 22 above)
Comment (SaarYogev) [reply src/pid/CtreController.java:77]: From what I can tell the default is zero for most of them, that's what will happen if you won't call them.
A lot of ctors is not a reason to make to move to public functions and mutable objects.
I agree that it's annoying that Java doesn't have default function parameters, but you can just use a [builder](https://en.wikipedia.org/wiki/Builder_pattern#Examples).

File: src/pid/CtrePIDController.java (hunk 24 above)
Comment (SaarYogev) [reply src/pid/CtrePIDController.java:40]: Both should move together, and `tolerance` should still be set in ctor.
If you use a builder you can use two `setTolerance` methods, one that accepts a single tolerance and one that accepts both 
`aboveTolerance` and `belowTolerance`.

Also, look at the Markdown Turorial in the bottom right of the comment textbox
image
to see how to make `variables` and [links](https://guides.github.com/features/mastering-markdown/)

Comment (SaarYogev) [reply src/pid/CtrePIDController.java:40]: Talking about tolerance and `isOnTarget`, I now read the previous reviews and saw [this comment I made](https://github.com/OnyxTronix2231/Lib-FRC/pull/18#discussion_r422470682), which is still accurate.
We do need to extract the implementation to a different class, to be able to create others in the future (like the stopping one we usually code-hacked with CTRE by waiting).

File: src/pid/PIDRunner.java (hunk 9 above)
Comment (SaarYogev) [reply src/pid/PIDRunner.java]: @AE173 I think we didn't include that in the new implementation, make sure it's in.

**Review by SaarYogev — CHANGES_REQUESTED** [review]
Also resolved what's addressed. The are still left for you to solve @AE173 .

File: src/pid/AbstractController.java (hunk 20 above)
Comment (AE173) [reply src/pid/AbstractController.java:9]: Done

File: src/pid/CtrePIDController.java [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+
+
+  @Override
+  public double getProcessVariable() {
+    if (this.ctreMotorController.getControlMode() == ControlMode.Velocity) {
+      return this.ctreMotorController.getSelectedSensorVelocity(pidSlot);
+    } else if (this.ctreMotorController.getControlMode() == ControlMode.Position) {
+      return this.ctreMotorController.getSelectedSensorPosition(pidSlot);
+    }
+    return -1;
```
Comment (AE173) [inline src/pid/CtrePIDController.java:33]: Throw exception instead

File: src/pid/CtrePIDController.java [hunk 30]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  public void restartControllerState() {
+    this.ctreMotorController.setIntegralAccumulator(0,
+        this.pidSlot, this.timeoutMs);
+  }
+
+  @Override
+  public void enable(PIDControlMode controlMode) {
+    if (controlMode == PIDControlMode.Position) {
```
Comment (AE173) [inline src/pid/CtrePIDController.java:56]: Update PID Values before enable

File: src/exceptions/UnsupportedControlModeException.java [hunk 31]
```diff
@@ -0,0 +1,8 @@
+package exceptions;
+
+public class UnsupportedControlModeException extends Exception {
```
Comment (nimrod46) [inline src/exceptions/UnsupportedControlModeException.java:3]: Use [unchecked exceptions](https://howtodoinjava.com/java/exception-handling/checked-vs-unchecked-exceptions-in-java/)

**Review by nimrod46 — CHANGES_REQUESTED** [review]

File: src/pid/CtrePIDController.java [hunk 32]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  public CtrePIDController(IMotorControllerEnhanced ctreMotorController, CtreEncoder ctreEncoder,
+                           PIDFTerms pidfTerms, int pidSlot, int timeoutMs) {
+    super(ctreMotorController, ctreEncoder, pidfTerms, pidSlot, timeoutMs);
+  }
+
+
+  @Override
+  public double getProcessVariable() throws UnsupportedControlModeException {
```
Comment (nimrod46) [inline src/pid/CtrePIDController.java:28]: Remove `throws` from the method signature

**Review by nimrod46 — CHANGES_REQUESTED** [review]

File: src/exceptions/UnsupportedControlModeException.java [hunk 33]
```diff
@@ -0,0 +1,8 @@
+package exceptions;
+
+public class UnsupportedControlModeException extends RuntimeException {
+  public UnsupportedControlModeException(){
+    super("This control mode is not supported." +
```
Comment (SaarYogev) [inline src/exceptions/UnsupportedControlModeException.java:5]: Add which ControlMode they were on, will make it easier to debug.

**Review by nimrod46 — APPROVED** [review]
